[Performance] Track waiting time when app is launched (app startup)

rachelmcr opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #9891
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
We want to improve the loading time on app startup (when the app is launched). To measure any performance improvements, we are adding a new event:

* `application_opened_waiting_time_loaded` (prop: `waiting_time`) (Event registration: 1664-gh-tracks-events-registration)

The `waiting_time` property reflects the time it takes for all of the requested data to be retrieved when the app is first launched and the user is logged in.

### Changes:

* Adds `application_opened_waiting_time_loaded` to `WooAnalyticsStat`, and tracks it in `waitingFinished(scenario:elapsedTime:)`.
* Adds a new waiting time tracker `AppStartupWaitingTimeTracker` to track the app startup time:
   * The tracker has a list of notifications to observe (`startupActionsToObserve`). These are all actions (primarily API requests) that have been identified as occurring on app startup. Internal ref: pe5pgL-32Y-p2
   * When the tracker is initialized, it initializes `WaitingTimeTracker` with the `appStartup` scenario and starts listening for notifications in `startupActionsToObserve`.
   * It has a static method `notify(action:withStatus:)` that is used around the app to send a notification that a specific startup action has been started or completed.
   * When an action is started, it's added to a list of pending actions. This is because not all of the possible startup actions are done every time the app is launched, so we want to know which ones are pending for a given app launch.
   * When an action is ended, it's removed from the list of pending actions. If no more actions are pending, we stop listening for notifications and trigger the analytics event.
* Starts the waiting time tracker when the app is being launched.
* Adds notifications where the startup actions are started/completed throughout the app.
* Updates `ABTest` to only refresh experiment assignments if `ExPlat.shared != nil`. Otherwise, we call `ExPlat.shared?.refresh` but it never completes (which would cause `fetchExperimentAssignments` to start in the waiting time tracker but never end).
* Updates `DefaultStoresManager.synchronizeEntities()` to bail if the app is authenticated without WPCom. Otherwise, we try to sync the WPCom account, account settings, and sites, but the syncing never completes because there's no `AccountStore` to dispatch the actions (which would cause `synchronizeEntities` to start in the waiting time tracker but never end).

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

### Logged out

1. Start logged out of the app.
2. Build and run the app.
3. When the app launches, confirm the event `application_opened_waiting_time_loaded` is **NOT** tracked.

### Logged in with WordPress.com account / Jetpack-connected store

1. Start logged in to the app with a WordPress.com account (with Jetpack-connected store).
2. Build and run the app.
3. When the app launches, confirm the event `application_opened_waiting_time_loaded` is tracked with the custom property `waiting_time` reflecting the startup waiting time in seconds.

### Logged in with the REST API

1. Start logged in to the app with the REST API (without a WordPress.com account / with a store not connected via Jetpack)
2. Build and run the app.
3. When the app launches, confirm the event `application_opened_waiting_time_loaded` is tracked with the custom property `waiting_time` reflecting the startup waiting time in seconds.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

rachelmcr committed:
Add a tracker to track the app startup waiting time

rachelmcr committed:
Use notification center to track when app startup actions are completed

rachelmcr committed:
Trigger notifications when setup requests are completed

rachelmcr committed:
Define pending actions separately from defining what notifications to observe

rachelmcr committed:
Move app startup notifications into synchronizeEntities

rachelmcr committed:
Only refresh assignments if there is a shared Explat instance

rachelmcr committed:
Only track the app startup waiting time when logged in

rachelmcr committed:
Don't try to sync account entities when authenticated without WPCom

rachelmcr committed:
Track app launch time

rachelmcr committed:
Trigger analytics event for app startup waiting time

rachelmcr committed:
Use helper to notify when startup action starts or ends

rachelmcr committed:
Add unit tests for AppStartupWaitingTimeTracker

rachelmcr committed:
Add documentation

Ecarrion commented on Experiments/Experiments/ABTest.swift:
```diff
@@ -52,14 +52,16 @@ public extension ABTest {
             let experimentNames = experiments.map { $0.rawValue }
             ExPlat.shared?.register(experiments: experimentNames)
 
+            NotificationCenter.default.post(name: .fetchExperimentAssignments, object: "started")
```
Rather than posting the notifications here with untyped objects, can we use the typed invocation on `WooAnalytics.refreshUserData()` where the `ABTest.start()` method is called?

Ecarrion commented on WooCommerce/Classes/Analytics/AppStartupWaitingTimeTracker.swift:
```diff
@@ -0,0 +1,111 @@
+        startupActionsPending.removeAll { $0 == startupAction }
+
+        // End the waiting time tracker when no more actions are pending
+        if startupActionsPending.isEmpty {
+            notificationCenter.removeObserver(self) // Stop listening to any notifications
+            guard stores.isAuthenticated else { // Don't track the waiting time if the user is logged out
+                return
+            }
+            super.end() // Calculate the elapsed time and trigger analytics event
```
Is it possible to have this scenario?

```
- Notification1-start
- Notification1-end
- Slight delay
- waiting time reported
- Notification2-start
- Notification2-end
```
In that scenario, we will get an inaccurate timer. Since we post notifications from different parts of the App, I'm not sure if we can control that that doesn't happen, can we?

Ecarrion reviewed (commented):
Do we care about API down / Not Internet connection scenarios? I wonder how that would affect the `waiting_time` metric analysis 🤔 

Also, an open question: Does sentry performance monitoring measure "somehow" the time the app takes to be functional? 

I'm trying to think if there is a better way to measure the perceived start-up time, that may be less prone to error scenarios 🤔 

Do you think measuring the time it takes from startup until the dashboard is loaded would be enough? Maybe there it would be easy to discard the API errors / No Internet events.

rachelmcr commented:
> Do we care about API down / Not Internet connection scenarios? I wonder how that would affect the waiting_time metric analysis 🤔

That's a good point. I was thinking that it's worth tracking waiting time even if there are failures, because even if there are backend errors it contributes to perceived loading time. But no internet connection scenarios (and any other immediate failures) could really skew the analysis.

> Also, an open question: Does sentry performance monitoring measure "somehow" the time the app takes to be functional?

Sort of. From what I've read it does track app startup time, but only measuring the time from app launch to when the UI appears. From [the docs](https://docs.sentry.io/platforms/apple/guides/ios/performance/instrumentation/automatic-instrumentation/#app-start-tracing):

> The SDK uses the process start time as the beginning of the app start and the [UIWindowDidBecomeVisibleNotification](https://developer.apple.com/documentation/uikit/uiwindowdidbecomevisiblenotification) as the end.

So performance monitoring could provide some useful data but it won't measure the same thing we're looking at here.

> Do you think measuring the time it takes from startup until the dashboard is loaded would be enough? Maybe there it would be easy to discard the API errors / No Internet events.

Thanks for bringing this up — this might be enough and could simplify things quite a bit (tracking fewer things and less possibility that the waiting time will be reported too early). I'll set this PR to draft and look into that approach.

rachelmcr commented:
@Ecarrion I started over with a different approach in https://github.com/woocommerce/woocommerce-ios/pull/9987, which I think addresses the concerns here and is generally a simpler and more straightforward approach.
